python/phase-1.md

```diff
@@ -0,0 +1,460 @@
+        workflow_id: str,
+        *,
+        run_id: Optional[str] = None
+        # ...
+    ) -> None:
+        pass
+
+    # This is not memoized, each invocation calls the server
+    async def result(self) -> T:
+        pass
+
+    async def cancel(self):
+        pass
+
+    async def describe(self) -> temporalio.WorkflowExecution:
+        pass
+
+    async def query(self, name: str, *args: Any) -> Any:
+        pass
+
+    async def signal(self, name: str, *args: Any):
+        pass
+
+    async def terminate(self, *, reason: Optional[str] = None):
+        pass
+```
+
+Notes:
+
+* This API supports `async with` or just `Client.connect` + `client.close()`, user's preference
+* We have chosen to, at first, only use Core's gRPC layer
+  * The generated API/gRPC code for the pure Python gRPC client is still available for use, just not by this wrapper
+* We are intentionally defining a `Client` instead of a `WorkflowClient` since this will be general purpose
+  * Having a general purpose client keeps things simple, but this will probably only ever contain the pieces above (and
+    overloads thereof) and async activity heartbeat/complete
+* We call it `start_workflow` instead of just `start` to be clearer
+* We have an `execute_workflow` to combine start + result waiting just as a shortcut
+* We have embedded `start_signal` and `start_signal_args` as parameters, so if the former is present we will perform a
+  signal with start. After discussion, it was determined this is acceptable.
+* After discussion, we determined that the concept of a "default task queue" at the client level is not worth it right
+  now.
+
+### Worker
+
+Exported from `temporalio` (but probably in `temporalio/worker/worker.py`, exported from `temporalio.worker` then
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        client: temporalio.Client,
+        *,
+        task_queue: str,
+        # Must be a mapping with the key as the activity name and the value as a
+        # callable to be called when the activity is invoked.
+        activities: Mapping[str, Callable[P, T]] = {},
+        # An executor to be used for synchronous activities. If set, the worker
+        # does not shut this down. If unset, the worker may use a thread pool
+        # internally.
+        activity_executor: Optional[concurrent.futures.Executor] = None,
+        # Same as activity_executor but for asynchronous activities.
+        async_activity_executor: Optional[concurrent.futures.Executor] = None,
+        # ...
+    ) -> None:
+        pass
+
+    # Just start + shutdown on finally (so can be cancelled via asyncio cancel)
+    async def run(self) -> None:
+        pass
+
+    async def start(self) -> None:
+        pass
+
+    async def shutdown(self) -> None:
+        pass
+
+    async def __aenter__(self) -> Worker:
+        await self.start()
+        return self
+
+    async def __aexit__(self) -> None:
+        await self.shutdown()
```

Comment:
I don't think this is required, what is this supposed to block on?

```
async with worker:
  # what goes here?
```

Reply:
```
async with worker:
  # everything you want to do with a worker
```

If you want to use it that way. The point is that it stops when done there. If you don't have your code encapsulated like that, you can just run the worker, you don't have to use `async with`.